File: packages/core/src/http/request.ts
import { ProviderError, classifyStatus } from "../errors.js";

export type FetchLike = (url: string, init?: RequestInit) => Promise<Response>;

export interface FetchOptions {
  timeoutMs: number;
  signal?: AbortSignal;
  /**
   * Streaming responses: keep the caller-abort relay attached after the
   * headers arrive so aborting mid-body still cancels the connection.
   * Without this flag the relay is detached as soon as fetch resolves
   * (right for one-shot JSON bodies; wrong for SSE streams).
   */
  streaming?: boolean;
}

/**
 * fetch with a per-attempt timeout and optional outer cancellation.
 * Network failures propagate as thrown errors; adapters wrap them into
 * ProviderError(kind: "network" | "timeout").
 */
export async function fetchWithTimeout(
  fetchImpl: FetchLike,
  url: string,
  init: RequestInit,
  opts: FetchOptions,
): Promise<Response> {
  const controller = new AbortController();
  const timer = setTimeout(() => {
    controller.abort(new DOMException("request timed out", "TimeoutError"));
  }, opts.timeoutMs);

  const onOuterAbort = () => controller.abort(opts.signal?.reason);
  if (opts.signal) {
    if (opts.signal.aborted) {
      clearTimeout(timer);
      throw new DOMException("aborted", "AbortError");
    }
    opts.signal.addEventListener("abort", onOuterAbort, { once: true });
  }

  try {
    return await fetchImpl(url, { ...init, signal: controller.signal });
  } finally {
    clearTimeout(timer);
    // For streams the relay stays attached for the body's lifetime; the
    // request's own completion/GC releases it.
    if (!opts.streaming && opts.signal) opts.signal.removeEventListener("abort", onOuterAbort);
  }
}

/**
 * Wrap a low-level failure into a retryable ProviderError.
 * Caller-cancellation AbortErrors are returned unwrapped so the engine can
 * distinguish "user cancelled" (never retry, never fall back) from
 * transport failures. Timeout aborts keep the TimeoutError name.
 */
export function toNetworkError(provider: string, err: unknown): ProviderError | DOMException {
  if (err instanceof ProviderError) return err;
  if (err instanceof DOMException && err.name === "AbortError") return err;
  const isTimeout = err instanceof DOMException && err.name === "TimeoutError";
  return new ProviderError(provider, isTimeout ? "timeout" : "network", errorMessage(err), {
    cause: err,
  });
}

/** True when the error is a caller-cancellation abort (never retryable). */
export function isAbortError(err: unknown): err is DOMException {
  return err instanceof DOMException && err.name === "AbortError";
}

/** Read Retry-After (seconds or HTTP-date) as milliseconds. */
export function parseRetryAfter(headers: Headers): number | undefined {
  const raw = headers.get("retry-after");
  if (raw === null || raw.length === 0) return undefined;
  const seconds = Number(raw);
  if (Number.isFinite(seconds)) return Math.max(0, Math.round(seconds * 1000));
  const date = Date.parse(raw);
  if (!Number.isNaN(date)) return Math.max(0, date - Date.now());
  return undefined;
}

export function errorMessage(err: unknown): string {
  if (err instanceof Error) return err.message;
  return String(err);
}

/**
 * Shared non-2xx handling. OpenAI and Anthropic error bodies both carry
 * { error: { message } }, so one implementation serves every adapter.
 */
export async function requireOk(resp: Response, provider: string): Promise<void> {
  if (resp.ok) return;
  const kind = classifyStatus(resp.status);
  let body: unknown;
  let message = `${provider}: HTTP ${resp.status}`;
  const text = await resp.text().catch(() => undefined);
  if (text !== undefined) {
    try {
      body = JSON.parse(text) as unknown;
      const errMsg = (body as { error?: { message?: string } }).error?.message;
      if (errMsg) message = `${provider}: ${errMsg}`;
    } catch {
      // non-JSON error body; keep the generic message
    }
  }
  throw new ProviderError(provider, kind, message, {
    status: resp.status,
    retryAfterMs: parseRetryAfter(resp.headers),
    body,
  });
}

File: packages/python/src/ai_router/http/request.py
"""HTTP utilities, timeout handling, and Retry-After parsing."""

from __future__ import annotations

import asyncio
import email.utils
import json
import time
import urllib.error
import urllib.request
from collections.abc import Callable, Coroutine, Mapping
from dataclasses import dataclass
from typing import Any

from ..errors import ErrorKind, ProviderError, classify_status


class AbortError(Exception):
    """Raised when an operation is cancelled by the caller."""


class TimeoutError(Exception):
    """Raised when an operation times out."""


@dataclass
class HttpResponse:
    status: int
    headers: dict[str, str]
    body: bytes
    text: str | None = None

    @property
    def body_text(self) -> str:
        if self.text is not None:
            return self.text
        return self.body.decode("utf-8", errors="replace")


FetchLike = Callable[[str, dict[str, str], bytes | None, int], Coroutine[Any, Any, HttpResponse]]


def _sync_fetch(
    url: str, headers: dict[str, str], body: bytes | None, timeout_ms: int, method: str = "POST"
) -> HttpResponse:
    req = urllib.request.Request(url, data=body, headers=headers, method=method if body is not None else "GET")
    try:
        with urllib.request.urlopen(req, timeout=timeout_ms / 1000.0) as resp:
            return HttpResponse(
                status=resp.status,
                headers=dict(resp.headers),
                body=resp.read(),
            )
    except urllib.error.HTTPError as e:
        return HttpResponse(
            status=e.code,
            headers=dict(e.headers),
            body=e.read(),
        )


async def default_fetch(url: str, headers: dict[str, str], body: bytes | None, timeout_ms: int) -> HttpResponse:
    return await asyncio.to_thread(_sync_fetch, url, headers, body, timeout_ms)


def parse_retry_after(headers: Mapping[str, str] | None) -> int | None:
    """Read Retry-After header (seconds or HTTP-date) as milliseconds."""
    if not headers:
        return None
    raw = None
    for k, v in headers.items():
        if k.lower() == "retry-after":
            raw = v.strip()
            break
    if not raw:
        return None

    try:
        seconds = float(raw)
        return max(0, int(round(seconds * 1000)))
    except ValueError:
        pass

    try:
        timetuple = email.utils.parsedate_tz(raw)
        if timetuple is not None:
            ts = email.utils.mktime_tz(timetuple)
            return max(0, int(round((ts - time.time()) * 1000)))
    except Exception:
        pass

    return None


def to_network_error(provider: str, err: Exception) -> Exception:
    """Wrap a low-level failure into a retryable ProviderError."""
    if isinstance(err, (ProviderError, AbortError, asyncio.CancelledError)):
        return err
    is_timeout = isinstance(err, (TimeoutError, asyncio.TimeoutError))
    return ProviderError(
        provider,
        "timeout" if is_timeout else "network",
        str(err) if str(err) else type(err).__name__,
        cause=err,
    )


def is_abort_error(err: Exception) -> bool:
    """True when the error is a caller-cancellation abort (never retryable)."""
    return isinstance(err, (AbortError, asyncio.CancelledError))


def require_ok(status: int, headers: Mapping[str, str] | None, body_text: str | None, provider: str) -> None:
    """Shared non-2xx status handling and error classification."""
    if 200 <= status < 300:
        return

    kind: ErrorKind = classify_status(status)
    body: Any = None
    message = f"{provider}: HTTP {status}"
    if body_text:
        try:
            body = json.loads(body_text)
            if isinstance(body, dict):
                err_obj = body.get("error")
                if isinstance(err_obj, dict) and "message" in err_obj:
                    message = f"{provider}: {err_obj['message']}"
                elif "message" in body:
                    message = f"{provider}: {body['message']}"
        except Exception:
            pass

    retry_after_ms = parse_retry_after(headers)
    raise ProviderError(
        provider=provider,
        kind=kind,
        message=message,
        status=status,
        retry_after_ms=retry_after_ms,
        body=body,
    )
